File: internal/console/ui/app/screens/presence.js
/* ==========================================================================
   Who is here — the only screen in this console that is about right now.

   EVERY OTHER SCREEN HERE IS A PHOTOGRAPH OR A FILM OF THE PAST. The funnel is
   everybody at once, cohorts are intakes followed forward, the item analysis is
   what a nightly job wrote. This one is worth nothing five minutes late, which
   is why it refreshes itself and why it says how old the number it is showing
   already is.

   # IT COUNTS AND IT DOES NOT LIST

   K-22 was amended and `Personal data` lists people now, under four conditions
   — bounded, minimal, counted, named. This screen would fail two, and the
   decisive one is COUNTED: it refreshes on a timer, so there is no moment at
   which anybody asked for it and nothing to record a purpose against. A roll of
   who is online, arriving every few seconds unbidden, is the browse with the
   whole of what makes a listing defensible missing.

   So the answer has no names in it, and this screen has nothing to draw them
   with even if one arrived. The number is the whole report, and an operator who
   needs a name has one already — or has the listing, which asks what they are
   looking for and writes down that they asked.

   # THE WINDOW IS NOT WRITTEN IN THIS FILE

   "Seen in the last five minutes" is what the number MEANS, and it comes back
   from the server beside it (K-16), for the same reason the item analysis's
   thresholds do: an interface holding its own copy of a definition keeps
   describing the old one for as long as nobody notices.

   # WHY THE COUNTS DO NOT ADD UP, AND WHY THAT IS NOT A BUG

   Somebody studying in two schools is present in both and is one person on the
   platform. So the total is its own number rather than a sum, and the screen
   says so where somebody would otherwise reach for a calculator.
   ========================================================================== */

import { esc } from '../dom.js';
import { get } from '../request.js';
import { txt, language } from '../../assets/language.js';

/* HOW OFTEN THIS ASKS AGAIN. It is deliberately slower than the heartbeat it
   watches: the server writes a session's last-seen at most once a minute, so
   asking every five seconds would draw the same number four times and cost
   twelve reads a minute for it. Fifteen seconds is fast enough that the screen
   feels live and slow enough that it is not lying about its own resolution. */
const EVERY = 15000;

export default async function presence(section) {
  const el = document.createElement('div');
  el.className = 'view';

  el.innerHTML =
    '<header class="view-head">' +
      '<span class="eyebrow mono">' + esc(txt('Measure')) + '</span>' +
      '<h1>' + esc(txt('Who is here')) + '</h1>' +
      '<p>' + esc(txt('People signed in and seen a moment ago, by school. This is the one number in the console that comes from the sessions table rather than from the event stream — presence is the question where being overwritten is the point, because nobody asks who was online last March.')) + '</p>' +
    '</header>' +

    /* NOTHING HERE IS A LIVE REGION, and that is a decision rather than an
       oversight. Every other screen in this console marks its answer
       `aria-live="polite"` because it redraws when somebody asks it to; this one
       redraws every fifteen seconds whether anybody asked or not, and a polite
       region would read the whole report out four times a minute, over the top
       of whatever was being read.

       The freshness line is not one either, for the same reason and less
       obviously: its timestamp changes on every pass, so `role="status"` there
       would announce a new clock reading every fifteen seconds — which is the
       same interruption wearing a smaller sentence. What announces this screen
       is the router naming it on arrival, once. */
    '<div id="body"><p class="checking">' + esc(txt('Reading…')) + '</p></div>';

  const body = el.querySelector('#body');

  let timer = null;
  await draw();
  timer = setInterval(draw, EVERY);

  /* THE TIMER STOPS WHEN THE SCREEN DOES. Without this, leaving for the funnel
     leaves a request running every fifteen seconds for as long as the tab is
     open — invisible, and exactly the kind of thing that is discovered in a
     log six months later. */
  return { title: section.name, el, onLeave: () => clearInterval(timer) };

  async function draw() {
    let answer;
    try {
      answer = await get('/console/api/v1/watch/presence');
    } catch (e) {
      /* A READ THAT FAILED IS NOT AN EMPTY PLATFORM. Nobody being here is a
         real answer that looks exactly like this, so the failure has to replace
         the numbers rather than be shown as zero people. */
      body.innerHTML = '<section class="block"><p class="none">' + esc(txt(e.message)) +
        '</p></section>';
      return;
    }

    const schools = answer.schools || [];
    const minutes = Math.round((answer.window_seconds || 0) / 60);
    const cadence = Math.round((answer.cadence_seconds || 0) / 60);

    body.innerHTML =
      '<section class="block">' +
        '<div class="block-top"><h2>' + esc(txt('On the platform')) + '</h2></div>' +
        '<span class="here-total">' + (answer.everywhere || 0) + '</span>' +
        /* TWO NUMBERS IN ONE SENTENCE, AND BOTH ARE HOLES. It was assembled —
           a window, a plural `s` chosen by hand, a cadence, another plural —
           which is the "1 trilhas" shape twice in one paragraph. Four cases,
           because both numbers can be one and each of the two languages
           agrees differently. */
        '<p class="aside">' +
          (minutes === 1 ? txt('Seen in the last minute.') : txt('Seen in the last %w minutes.').replace('%w', minutes)) +
          ' ' +
          (cadence === 1 ? txt('A session says it is still in use at most once a minute, so this number is accurate to that and no better — it moves in steps rather than smoothly, and that is the heartbeat rather than the platform.') : txt('A session says it is still in use at most once every %c minutes, so this number is accurate to that and no better — it moves in steps rather than smoothly, and that is the heartbeat rather than the platform.').replace('%c', cadence)) +
        '</p>' +
      '</section>' +

      '<section class="block">' +
        '<div class="block-top"><h2>' + esc(txt('By school')) + '</h2></div>' +
        (schools.length === 0
          ? '<p class="none">' + esc(txt('There are no schools on this platform yet.')) + '</p>'
          : '<div class="here-all">' + schools.map(tile).join('') + '</div>') +

        /* THE TWO NUMBERS DISAGREE ON PURPOSE, said once, here, rather than
           discovered by somebody adding the tiles up. */
        (schools.length > 1
          ? '<p class="aside">' + txt('These do not add up to the number above, and should not: somebody studying in two schools is present in both and is <b>one person</b> on the platform.') + '</p>'
          : '') +
      '</section>' +

      '<section class="block">' +
        '<div class="block-top"><h2>' + esc(txt('What this does not count')) + '</h2></div>' +
        '<p class="aside">' + esc(txt(answer.not_counted || '')) + '</p>' +

        /* THE RULE ON THE SCREEN, because "why can I not see who" is a question
           an operator will have, and the answer is a rule rather than an
           omission somebody can be asked to fix.

           IT IS A SHARPER ANSWER SINCE K-22 WAS AMENDED, not a weaker one:
           there IS a place to look people up, and the reason it is not this one
           is that this page arrives on a timer with nobody having asked. */
        '<p class="aside">' + txt('And it does not say <b>who</b>. This page refreshes on its own, so there is no moment where somebody asked for it and nothing to record a reason against — which is what a list of people has to carry. Looking somebody up is <b>Personal data</b>, where every page is recorded with what was searched for.') + '</p>' +
      '</section>' +

      '<p class="freshness mono">' +
        esc(txt('Read at %t · refreshing every %n s')
          .replace('%t', clock(answer.as_of))
          .replace('%n', Math.round(EVERY / 1000))) + '</p>';
  }
}

/* One school, and the number is the tile rather than a line in a table. A
   table of two columns and four rows is a table; this is four numbers somebody
   glances at. */
function tile(school) {
  return '<div class="here' + (school.people > 0 ? ' here-live' : '') + '">' +
    '<span class="here-n mono">' + (school.people || 0) + '</span>' +
    '<span class="here-school">' + esc(school.name || '') + '</span>' +
    '<span class="here-slug mono">' + esc(school.slug || '') + '</span>' +
  '</div>';
}

/* The time the server answered, in the reader's own clock. A count of people
   present is worthless without knowing when it was counted, and a screen that
   silently stopped refreshing looks identical to a quiet Tuesday. */
function clock(when) {
  const at = new Date(when);
  if (Number.isNaN(at.getTime())) return txt('an unknown moment');
  /* THE READER'S OWN CLOCK, IN THE READER'S OWN LANGUAGE. `toLocaleTimeString()`
     with no argument is the BROWSER's locale, which is the defect the dates and
     the money on the other screens had: a 12-hour clock with `PM` on a screen
     otherwise in Portuguese. */
  return at.toLocaleTimeString(language() === 'pt' ? 'pt-BR' : 'en-GB');
}

File: internal/identity/presence.go
package identity

import (
	"context"
	"fmt"
	"time"

	"github.com/codeschool-ing/schooling/internal/platform/setting"
	"github.com/google/uuid"
)

/* Who is here right now.

   # THE ONLY NUMBER IN THE CONSOLE THAT IS NOT FROM THE EVENT STREAM

   K-03 says statistics come from `events` and never from current state, because
   current state has been overwritten. This is the one question where being
   overwritten is the point. Nobody asks who was online last March — and the
   stream could not answer it if they did, because answering would need an event
   for LEAVING and no browser reliably sends one. A tab closed by a flat battery
   emits nothing, so "still here" is inferred from an absence either way; doing
   it from a column is the same arithmetic two tables closer to the fact.

   # IT COUNTS PEOPLE AND NOT SESSIONS

   Same rule as the funnel's, for the same reason: somebody with a laptop and a
   phone is one person, and counting sessions would make the number go up when
   the population did not. So the count is DISTINCT over accounts, per school,
   and the platform-wide figure is distinct again across schools rather than the
   sum of them — a person studying in two schools is in both numbers and once in
   the total.

   # AND IT DOES NOT SAY WHO THEY ARE

   K-22 no longer says people are never listed — it was amended, and a listing
   is allowed where it is bounded, minimal, counted and named. This would be
   none of those, and the last two are what settle it: a list of who is online
   is refreshed on a timer, so there is no moment at which somebody asked for it
   and nothing to record against a purpose. That is a browse of the population
   with the one thing the amendment rests on missing.

   So this answers a count and has no shape in which to return a name. The
   operator who needs one has it already, because somebody wrote in — and the
   operator who does not have it has `Look`, which records that they asked.
*/

/*
PresenceWindow is how recently somebody has to have been seen to be here.

	IT WAS A CONSTANT, AND THE ARGUMENT FOR THAT WAS RIGHT ABOUT ONE END ONLY.
	It read: the answer follows from the heartbeat, because `Verify` writes at
	most once a minute, so a window narrower than that reports people as gone
	between their own requests — and a window much wider stops being "now".

	The first half is a fact and the second half is a preference wearing its
	clothes. Nothing measures where "now" ends. Three minutes and ten minutes
	are both defensible readings of the same screen, and which one an operator
	wants depends on what they are watching for: a deploy going wrong wants the
	short end, an evening's traffic wants the long one.

	SO THE FACT BECAME THE FENCE AND THE PREFERENCE BECAME THE KNOB. `Least` is
	twice the cadence, which is the constraint the old comment actually
	established — `TestTheWindowCannotBeSetInsideTheHeartbeat` holds the two
	together, because the day somebody changes the cadence is the day this
	bound stops being derived from anything.

	`Most` is half an hour, where a count of "who is here" has plainly become a
	count of who was here.

	IT STILL TRAVELS TO THE SCREEN BESIDE THE NUMBER IT PRODUCED (K-16) rather
	than being written down again there — and it matters more now, not less: a
	screen with the window hard-coded would label today's count with yesterday's
	span the moment somebody moved it.
*/
var PresenceWindow = setting.Declared{
	Name:     "identity.presencewindow",
	Unit:     setting.Minutes,
	Least:    2,
	Most:     30,
	Fallback: 5,
	Why: "how recently somebody has to have been seen to count as here. Nothing measures " +
		"where \"now\" ends — a deploy going wrong is watched at the short end and an " +
		"evening's traffic at the long one — so this is what the platform means by now. " +
		"It cannot go below twice the heartbeat, because a window narrower than the " +
		"writing cadence reports people as gone between their own requests.",
}

// Window is the declaration's answer as a duration, which is what everything
// that reads it actually wants. The registry counts in whole minutes because a
// screen and an audit entry both do; this is the one place that conversion
// lives, so it cannot be written down two ways.
func Window(minutes int) time.Duration {
	return time.Duration(minutes) * time.Minute
}

// PresenceCadence is what the heartbeat costs, sent alongside the window
// because it is the other half of what the number means: a count of people seen
// in the last five minutes, refreshed no oftener than once a minute per person,
// is accurate to a minute and no better. Saying so is cheaper than somebody
// discovering it from a graph that moves in steps.
const PresenceCadence = time.Minute

// Present is one school and how many people are in it.
type Present struct {
	School uuid.UUID
	People int
}

// Presence answers who is here, per school and in total.
//
// The two numbers do not add up on purpose — see the file header — so they come
// back from one query over one set of rows rather than from two reads that
// could disagree about the second somebody arrived.
func (s *Store) Presence(ctx context.Context, window time.Duration) ([]Present, int, error) {
	rows, err := s.pool.Query(ctx, `
		WITH here AS (
			SELECT DISTINCT s.last_seen_tenant AS school, s.account_id
			FROM sessions s
			JOIN accounts a ON a.id = s.account_id
			WHERE s.revoked_at IS NULL
			  AND s.expires_at > now()
			  AND s.last_seen_tenant IS NOT NULL
			  AND s.last_seen_at > now() - make_interval(secs => $1::double precision)

			  -- AN OPERATOR LOOKING AT A STUDENT'S SCREENS IS NOT THE STUDENT
			  -- BEING HERE. A viewing is a session on the school's host and
			  -- would otherwise count as one more person present — which is a
			  -- number that goes up when support gets busy and says nothing
			  -- about anybody studying.
			  AND s.viewed_by IS NULL

			  -- Seeded students hold no sessions at all: the seeder writes a
			  -- past into the event stream and never signs anybody in. So this
			  -- exclusion removes nothing today and is here because K-11 is
			  -- about what every aggregate does rather than about what happens
			  -- to be reachable — and it gets no switch beside it, because a
			  -- control whose two settings draw the same picture is a control
			  -- that teaches people not to trust the others.
			  AND NOT a.synthetic
		), per_school AS (
			SELECT school, count(*) AS people FROM here GROUP BY school
		)
		SELECT school, people, (SELECT count(DISTINCT account_id) FROM here)
		FROM per_school
		ORDER BY people DESC, school
	`, window.Seconds())
	if err != nil {
		return nil, 0, fmt.Errorf("identity: reading who is here: %w", err)
	}
	defer rows.Close()

	var out []Present
	var everywhere int
	for rows.Next() {
		var one Present
		if err := rows.Scan(&one.School, &one.People, &everywhere); err != nil {
			return nil, 0, fmt.Errorf("identity: reading who is here: %w", err)
		}
		out = append(out, one)
	}
	if err := rows.Err(); err != nil {
		return nil, 0, fmt.Errorf("identity: reading who is here: %w", err)
	}
	// No rows means nobody is anywhere, and `everywhere` was never scanned —
	// which is the zero it should be.
	return out, everywhere, nil
}
